pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -49,130 +53,343 @@ import (
 // These are intentionally pre-computed based on events such that lookups are efficient.
 type AmbientIndex struct {
 	mu sync.RWMutex
-	// byService indexes by Service (virtual) *IP address*. A given Service may have multiple IPs, thus
-	// multiple entries in the map. A given IP can have many workloads associated.
-	byService map[string][]*model.WorkloadInfo
-	// byPod indexes by Pod IP address.
-	byPod map[string]*model.WorkloadInfo
-
-	// Map of ServiceAccount -> IP
-	// TODO: currently, this is derived from pods. To be agnostic to the implementation,
-	// we should actually be looking at Gateway.status.addresses.
-	// This may be an external address (possibly even a DNS name we need to resolve), an arbitrary IP,
-	// or a reference to a service.
-	// If its a reference to a Service then we can find the underlying pods in that service, as an optimization.
-	waypoints map[model.WaypointScope]sets.String
-
-	// serviceVipIndex maintains an index of VIP -> Service
-	serviceVipIndex *kclient.Index[*v1.Service, string]
+
+	byIP map[string]*model.WorkloadInfo
+
+	// podIPIndex maintains an index of IP -> Pod
+	podIPIndex *kclient.Index[*v1.Pod, string]
+	// serviceVipIndex maintains an index of IP -> Pod
+	serviceVipIndex        *kclient.Index[*v1.Service, string]
+	sliceIPIndex           *kclient.Index[*discoveryv1.EndpointSlice, string]
+	waypointsIndex         *kclient.Index[*v1.Pod, model.WaypointScope]
+	services               kclient.Client[*v1.Service]
+	controller             *Controller
+	queue                  controllers.Queue
+	workloadEntriesIPIndex *kclient.Index[*config.Config, string]
 }
 
 // Lookup finds a given IP address.
 func (a *AmbientIndex) Lookup(ip string) []*model.WorkloadInfo {
 	a.mu.RLock()
 	defer a.mu.RUnlock()
 	// First look at pod...
-	if p, f := a.byPod[ip]; f {
+	if p, f := a.byIP[ip]; f {
 		return []*model.WorkloadInfo{p}
 	}
-	// Fallback to service. Note: these IP ranges should be non-overlapping
-	return a.byService[ip]
-}
+	// TODO: we need an index here
+	var workloads []*model.WorkloadInfo
+	for _, w := range a.byIP {
+		if _, f := w.VirtualIps[ip]; f {
+			workloads = append(workloads, w)
+		}
+	}
+	return workloads
+}
+
+func (a *AmbientIndex) Reconcile(name types.NamespacedName) error {
+	ip := name.Name
+	log := log.WithLabels("controller", "ambient", "ip", ip)
+	log.Debugf("reconcile")
+	// Should be 0 or 1 pod, unless we are in a weird state
+	pods := a.podIPIndex.Lookup(ip)
+	var pod *v1.Pod
+	if len(pods) > 0 {
+		// TODO: select newest
+		pod = pods[0]
+	}
+
+	// Should be 0 or 1 entry, unless we are in a weird state
+	entries := a.workloadEntriesIPIndex.Lookup(ip)
+	var entry *config.Config
+	if len(entries) > 0 {
+		entry = entries[0]
+	}
+
+	// We can have many slices (part of many services). Note each slice contains other IPs as well; we need to filter
+	epSlices := a.sliceIPIndex.Lookup(ip)
+	slices.SortFunc(epSlices, func(a, b *discoveryv1.EndpointSlice) bool {
```

Comment:
This seems potentially pretty expensive.. if we're hitting `Reconcile` for `n` IPs in a slice it's like `n^2`ish. 

Is it possible our indexer can pre-sort things?

Reply:
We should only have 1 slice in 99% of cases, this is only for the edge case where a single pod is in 2 different slices because it moved from one to the other